Rust Bindings for Key

PhilippGackstatter opened the pull request:
Due to #2839 growing too big, this PR only contains the Rust bindings for Key and is ready to be reviewed. The `elektra_sys` crate, that contains the raw bindings, as well as the CMake integration can also be reviewed. The Readme is not finished.

If you want to see what the documentation looks like, run `cargo doc` and then open `target/doc/elektra/index.html`.

## Basics

Check relevant points but **please do not remove entries**.
Do not describe the purpose of this PR in the PR description but:

- [x] Short descriptions should be in the release notes (added as entry in
      `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add something to the the release notes.**
- [ ] Longer descriptions should be in documentation or in design decisions.
- [x] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [x] I added unit tests
- [x] I ran all tests locally and everything went fine
- [x] affected documentation is fixed
- [x] I added code comments, logging, and assertions (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] meta data is updated (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))

## Review

Reviewers will usually check the following:

- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)
- [ ] Documentation is concise and good to read
- [ ] Examples are well chosen and understandable

## Labels

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the build server is happy** and also you
  say that everything is ready to be merged.

markus2330 commented on src/bindings/rust/README.md:
```diff
@@ -0,0 +1,59 @@
+Using the `buildelektra-stretch-full` docker image, we have to install cargo and rustc via [rustup](https://rustup.rs).
+We also need to install rusts formattings tools with `rustup component add rustfmt`.
+
+Bindings are generated when buildling the `elektra-sys` crate using `rust-bindgen`. The `build.rs` script in the `elektra-sys` crate calls and configures bindgen. It also emits additional configuration for `rustc` to tell it what library to link against, and where to find it.
```
```suggestion
Bindings are generated when building the `elektra-sys` crate using `rust-bindgen`. The `build.rs` script in the `elektra-sys` crate calls and configures bindgen. It also emits additional configuration for `rustc` to tell it what library to link against, and where to find it.
```

markus2330 commented on src/bindings/rust/README.md:
```diff
@@ -0,0 +1,59 @@
+We also need to install rusts formattings tools with `rustup component add rustfmt`.
+
+Bindings are generated when buildling the `elektra-sys` crate using `rust-bindgen`. The `build.rs` script in the `elektra-sys` crate calls and configures bindgen. It also emits additional configuration for `rustc` to tell it what library to link against, and where to find it.
+Bindgen expects a `wrapper.h` file that includes all headers that bindings should be generated for. Since these headers could have additional includes, the `build.rs` also passes the elektra source and binary directory to clang, such that these includes can be found.
```
This is already implementation detail? First explain the user-view of the Rust binding (How to use it.)

markus2330 commented on src/bindings/rust/README.md:
```diff
@@ -0,0 +1,59 @@
+#### KDB
+
+For improved readability, import all the enum variants directly. Through error nesting, you can catch specific errors and ignore other ones.
+TODO: Complete example
```
Start with minimal hello world or make line-by-line walk through.

markus2330 commented on src/bindings/rust/README.md:
````diff
@@ -0,0 +1,59 @@
+
+```rust
+use KDBError::*;
+use LogicalError::*;
````
put this into some hierarchy so that not so many use statements are needed?

PhilippGackstatter commented:
> Thank you for splitting up the PRs. Is this now the minimal PR?

I guess the actual minimal pr is #2826 with only cmake & elektra_sys and none of the safe wrappers. I was about to close that one, but if you prefer I can finish that first.

> Please finish the README as it is essential for review (I'll do as instructed in the README and try out if everything works.)
> 

I'll get on that.

markus2330 commented:
> I guess the actual minimal pr is #2826 with only cmake & elektra_sys and none of the safe wrappers. I was about to close that one, but if you prefer I can finish that first.

Yes, then finish this first!

PhilippGackstatter commented:
> Please finish the README as it is essential for review (I'll do as instructed in the README and try out if everything works.)

I've added an example using some common methods in the Readme.
 
> Please mark it as "ready to merge" once you finished everything here. Please also make sure that the build server is happy (all jobs pass).

I don't think I have the permission to set labels. I can leave a comment once it's ready.

PhilippGackstatter commented:
I think this PR is now ready for review. I tried to follow the advice in the [rust api guideline](https://rust-lang-nursery.github.io/api-guidelines/about.html), for instance not naming getters with the get prefix, so there's now `name()` instead of `get_name()`.
A couple of notes
- On they keys, there is a method `meta_iter`, that returns an iterator over the metakeys. I could call `keyRewindMeta` inside this method, so that users wouldn't have to remember to call that method. I'm wondering if that is a good idea.
- Methods that return strings, have to convert from the CStr to a utf-8 rust string. Ideally, users shouldn't have to deal with this error since it is rare but the methods are called often. So we can either panic on conversion errors, or use [`to_string_lossy`](https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.to_string_lossy) which would turn "Hello \xF0\x90\x80World\0" into "Hello �World". I currently have the latter implemented, because a conversion error is not severe enough to justify a panic. I think this is the best trade-off we can get for the majority of users.
- I did not provide an implementation of some methods, let me know if I should implement them anyway
  - keyNeedsSync (named in doc/todo/FUTURE)
  - keyValue (returns a void pointer)
  - keyGetNameSize & keyGetName (not needed since keyName exists)
  - keyGetUnescapedNameSize (integrated in name iterator)

markus2330 commented:
> On they keys, there is a method meta_iter, that returns an iterator over the metakeys. I could call keyRewindMeta inside this method, so that users wouldn't have to remember to call that method. I'm wondering if that is a good idea.

Only, if the keyset is a duplicate. I think the best user experience is if external iterators are actual external iterators and do not mess around with the object itself.

For meta-data this is currently possible by creating a new key, use keyCopyAllMeta and use the copied key for iteration. By stealing the reference to the internal meta-data keyset a more elegant implementation would be possible (identical to iterating over the keyset). elektraKeyGetMetaKeySet in kdbproposal.h would allow that.

> to_string_lossy/keyValue

It would be nice if it is somehow possible to get (binary) data without any conversion. It is a minor issue, though.

The other methods you can safely ignore for now.

markus2330 commented on src/bindings/rust/README.md:
```diff
@@ -27,9 +27,63 @@ fn main() {
+extern crate elektra;
+use elektra::{KeyBuilder, ReadableKey, StringKey, WriteableKey};
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
```
Can you put this example in a separate file (compiled with the build system)? Here in the tutorial there should be rather very small example codes with explanations.

PhilippGackstatter replied:
I shortened the example in the Readme to a very simple one and linked to the full example that is run together with `cargo test`, so is part of the build system.

markus2330 commented:
> I don't think I have the permission to set labels. I can leave a comment once it's ready.

Yes, please leave a comment and request for review.

PhilippGackstatter commented:
The build server seems to time out on `linux` or not move forward in `jenkins`.

> Yes, please leave a comment and request for review.

I'm otherwise done with this PR, so please review.

markus2330 commented:
> I'm otherwise done with this PR, so please review.

Perfect, I was just thinking about reviewing it.

markus2330 commented on src/bindings/rust/elektra/src/examples.rs:
```diff
@@ -0,0 +1,47 @@
+        assert_eq!(key.meta("pineapple")?.value(), "🍍");
+
+        // We can iterate over the metakeys
+        key.rewind_meta();
```
It would be nice if this is not needed (copy meta-data to other key) OR make the other iterators (like the name iterator) also like this.

PhilippGackstatter replied:
Yes that would be really nice! But I'm not sure what you mean with copying metadata? Basically copy the key, `rewind_meta()` on that key and then iterate over that? 
That would work, but for the KeySet later rewinding will still be necessary.

markus2330 replied:
Yes, duplicating the whole key (set) for the iterator also works.

> That would work, but for the KeySet later rewinding will still be necessary.

Why? If the iterator duplicates the KeySet it should be fine.

PhilippGackstatter replied:
Right, because duplicating a KeySet is only a shallow copy. But with KeySets I could also store the cursor, iterate over it and then restore it in the destructor.

PhilippGackstatter commented on src/bindings/rust/elektra/src/keybuilder.rs:
```diff
@@ -0,0 +1,114 @@
+    }
+}
+
+#[cfg(test)]
```
Isn't that enough? The tests are run on the build server every time right?

markus2330 replied:
There are already some tests that compile something during the tests (the external tests) so it is not a big problem as-is. But in general people expect that all compilation happens during `make` and `make test` only executes tests (and does not build them).

If it is difficult to realize and/or not the "Rust way", I am ok with how it is.

PhilippGackstatter replied:
I see what you mean. But hiding the tests behind that feature gate is the standard way of doing tests and `cargo test` therefore also compiles the tests before running them. So this is the "Rust way", afaik.

markus2330 replied:
Yes, then let leave it as-is.

PhilippGackstatter commented on src/bindings/rust/README.md:
```diff
@@ -8,6 +8,21 @@
 
+## Build
+
+To build the bindings explicitly as part of the elektra build process, we add the option `rust` to `-DBINDINGS`. After [building libelektra](../../../doc/COMPILE.md), we can now add the `elektra` crate to the dependencies. The exact paths depends on your system.
```
The current status is `- infos/status = experimental maintained`. I think it will be included by default if I remove experimental. Should I do that?

markus2330 replied:
I would be okay with that. You can also do it later when you are more confident.

PhilippGackstatter commented on src/bindings/rust/README.md:
```diff
@@ -8,6 +8,21 @@
+
+## Documentation
+
+Documentation can be built in the `src/bindings/rust/` directory, by running `cargo doc` and opening `target/doc/elektra/index.html`.
```
Once I push the crate to https://crates.io/, the documentation is automatically built and deployed to https://docs.rs/. That's where rust devs expect documentation to be and I think it's the easiest route.

I included a full example for key, later another for KeySet and also one for KDB (later PRs). Additionally, many methods have an example on how to use them. What else should I add?

markus2330 replied:
> That's where rust devs expect documentation to be and I think it's the easiest route.

Yes, sounds good!

> I included a full example for key

Thank you!

> What else should I add?

I was also missing infos about StringKeys vs. BinaryKeys. But maybe that gets clearer with the KeySet PR.

PhilippGackstatter replied:
> I was also missing infos about StringKeys vs. BinaryKeys. But maybe that gets clearer with the KeySet PR.

The Readme states the (small) differences between the two Key types under the Key section.

markus2330 replied:
Yes, I read this and the difference is clear. But how to actually use them in code (especially together with KeySets) was not described.

PhilippGackstatter replied:
I've added an example for `BinaryKey` and will add usage examples with KeySet in the KeySet-PR, if that's okay.

PhilippGackstatter commented on src/bindings/rust/README.md:
```diff
@@ -27,11 +42,34 @@ fn main() {
+
+Compared to the C-API, there are two distinct key types, `StringKey` and `BinaryKey`. With these, type mismatches such as calling `keyString` on a `BinaryKey` is not possible. The only difference between them is the type of value you can set and get from them. They are only wrappers over the `Key` from the C-API.
+
+The functionality of the keys is split into two traits, `ReadableKey` and `WritableKey`, which define methods that only read information from a key, and modify a key, respectively. For example, the method to retrieve metakeys only returns a key that implements `ReadableKey`, which is named `ReadOnly`. The keys returned cannot be modified in accordance to the design.
```
I did not read anything like that in the api documentation. Where do these occur?

markus2330 replied:
For a key that was appended to a KeySet you cannot change the name. I fixed the documentation 23a2f8f756a8c0fdbd980284b7a5ae4ca0802aa2

PhilippGackstatter replied:
Ok, I was confused because you said there are keys where you can only write to the value, but you probably meant reading?

The `set_name` method handles this by returning an error if the Name is read only (unfortunately this error is called KeyNameInvalidError, because that error can also occur, but both have the same error code).
For simplicity of the design, I decided not to prevent setting the name of a key in a keyset at the type system level.

markus2330 replied:
Ok, sorry for the confusion. Please also add this to the docu of set_name.

PhilippGackstatter commented on src/bindings/rust/README.md:
````diff
@@ -8,6 +8,21 @@
+elektra = { version = "0.9.0", path = "~/git/libelektra/build/src/bindings/rust/elektra" }
+# Directly depending on elektra-sys is only needed if you need to use the raw bindings
+elektra-sys = { version = "0.9.0", path = "~/git/libelektra/build/src/bindings/rust/elektra-sys" }
+```
````
I've added a full example description to the Readme.

PhilippGackstatter commented on src/bindings/rust/elektra/src/examples.rs:
```diff
@@ -0,0 +1,47 @@
+#[cfg(test)]
```
I've removed the file and created a full cargo project as an example instead, which will probably be more helpful.

PhilippGackstatter commented on doc/news/_preparation_next_release.md:
```diff
@@ -111,6 +111,7 @@ you up to date with the multi-language support provided by Elektra.
 ### Rust
 
 - Add the `elektra-sys` crate which contains raw bindings to libelektra for Rust. _(Philipp Gackstatter)_
+- Add the `elektra` crate which contains safe wrapper methods for the raw bindings. See the [Readme](../../src/bindings/rust/README.md) for more. _(Philipp Gackstatter)_
```
I've added some more details, let me know if that's what you had in mind.

markus2330 commented on src/bindings/rust/README.md:
```diff
@@ -8,30 +8,102 @@
+
 ## Usage
 
 ### Raw Bindings
```
I would move this to the end. So first key, keyset, kdb and then about raw bindings.

markus2330 commented:
Something went wrong in merging. Can you please rebase to libelektra's master?

PhilippGackstatter commented:
> Something went wrong in merging. Can you please rebase to libelektra's master?

I tried to rebase already, but on almost every commit there is a conflict, so it's quite painful to do. That's why I merged a commit that successfully built on master. Shouldn't that work just the same?

sanssecours commented:
> I tried to rebase already, but on almost every commit there is a conflict, so it's quite painful to do. 

I think one problem is that this PR contains already merged commits. Maybe you can do a interactive rebase on Elektra’s master branch:

```sh
git rebase -i master
```

. After you entered the command above your text editor will show a list of the relevant commits:

```
pick 668e5701c Rust: Add bindings gen & cmake integration
pick 96518dfa6 Rust: Add basic tests for bindings
…
pick 3ef2e0c0d Rust: Move Raw Bindings usage to the end
```

. I think you can drop the first seven commits, since they [were already merged](https://github.com/ElektraInitiative/libelektra/pull/2826). After that you still have to fix merge conflicts though. I tried it myself, but gave up after the second merge conflict, since I do not have the knowledge to fix the conflicts properly.

> That's why I merged a commit that successfully built on master. Shouldn't that work just the same?

If everything works out fine, the answer should be yes. However, it is very hard to review this PR, since GitHub shows a [lot of irrelevant changes in the file view](https://github.com/ElektraInitiative/libelektra/pull/2905/files).

markus2330 commented:
There seem to be formatting problems?

PhilippGackstatter commented:
I'm don't know why this failed. @sanssecours do you have an idea maybe?

sanssecours commented:
> I'm don't know why this failed. @sanssecours do you have an idea maybe?

This failure is caused by an issue of the Tcl plugin. We used the (old style) variable `Boost_INCLUDE_DIR` instead of `Boost_INCLUDE_DIRS` in the CMake code of the plugin. This problem was fixed in commit e20d0bd8. You should be able to fix the failing Travis build job, by rebasing this pull request on the `master` branch.

PhilippGackstatter commented:
> You should be able to fix the failing Travis build job, by rebasing this pull request on the `master` branch.

I've rebased to the latest master commit and the problem seems to be gone, but the link checker fails now. And the Apple GCC job fails as well.
The latest master build succeeded, so it seems like I've introduced something to cause it.

sanssecours commented:
> I've rebased to the latest master commit and the problem seems to be gone, but the link checker fails now.

The link checker also reports temporary unavailable URLs. Unfortunately our main Jenkins server (that hosts https://build.libelektra.org) is very underpowered. Consequently connections to https://build.libelektra.org will often time out. That is also what happen in the link checker build job according to the error output:

```
doc/BUILDSERVER.md:279:0 https://build.libelektra.org/jenkins/job/libelektra
doc/BUILDSERVER.md:293:0 https://build.libelektra.org/jenkins/job/libelektra
doc/BUILDSERVER.md:57:0 https://build.libelektra.org/jenkins/job/libelektra
doc/news/2016-06-14_0.8.17.md:120:0 https://build.libelektra.org/job/elektra-icc
doc/news/2014-11-04_0.8.9.md:110:0 https://build.libelektra.org/job/elektra-gcc-configure-debian
doc/news/2014-11-04_0.8.9.md:119:0 https://build.libelektra.org/job/elektra-gcc-configure-mingw-w64
doc/news/2014-11-04_0.8.9.md:107:0 https://build.libelektra.org/job/elektra-gcc-i386
doc/news/2018-03-27_0.8.22.md:170:0 https://build.libelektra.org/job/elektra-haskell
doc/news/2018-03-27_0.8.22.md:272:0 https://build.libelektra.org/job/elektra-haskell
doc/news/2014-06-21_0.8.6.md:15:0 https://build.libelektra.org/job/elektra-gcc47-all
doc/news/2016-06-14_0.8.17.md:118:0 https://build.libelektra.org/job/elektra-git-buildpackage-jessie
```

. I restarted the build job `🔗 Check`, which got rid of the temporary errors, but now shows an invalid link to https://debian-stretch-repo.libelektra.org. The cause of this is the (very unstable) server that hosts the Debian package repo (`a7`), which is currently down. Since `a7` also acts as a bridge to our other Jenkins hardware (`i7` and `v2`) Jenkins will not execute any build jobs, until someone restarts `a7`. For more information about these problems, please take a look at issue #160.

> And the Apple GCC job fails as well.

The build job only reported a temporary problem. It was unable to download some package. I restarted the build job.

> The latest master build succeeded, so it seems like I've introduced something to cause it.

I do not think so. As written above, the build failures are mainly a problem of our slow and/or unreliable Jenkins hardware.

markus2330 commented:
Thank you for the extensive explanation!
